Perform PowerOps for ComputerSystem

llvieira opened the pull request:
Closes #82.
Mapping Redfish power values to OneView power values.

To see how the mapping was made, consult the [table mapping](https://slack-files.com/T6B3FAKNK-F78NNLHM2-3f708824cf).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hewlettpackard/oneview-redfish-toolkit/88)
<!-- Reviewable:end -->

ffdarkpenguin commented:
Review status: 0 of 3 files reviewed at latest revision, 15 unresolved discussions.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 35 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv46Jok5qnmDzok2aKF:-Kv46JolLM1GE_ORoqHE:b-svz6ri) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L35)):*
> ```Python
> 
> computer_system = Blueprint(
>     "computer_system", __name__, url_prefix="/redfish/v1/Systems/")
> ```

I don't agre with this. We removed url_prefix from app registering process because it there were conflict. If someone else implement /Systems/Whatever in another place will fail because you took the hole /Systems to you.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 107 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv477SfYNIqZ0phW7tH:-Kv477SfYNIqZ0phW7tI:b-mtcrmv) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L107)):*
> ```Python
> 
> 
> @computer_system.route("<uuid>/Actions/ComputerSystem.Reset", methods=["POST"])
> ```

fix comment above and put the hole URI here. 

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 142 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv47ccPnmfQk_ECqSbi:-Kv47ccPnmfQk_ECqSbj:b-in8u57) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L142)):*
> ```Python
> 
>         # Build Computer System object and validates it
>         computer_system_class = ComputerSystem(sh, sht)
> ```

its an object no a class

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 148 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv47noWWBvfwNbmolZT:-Kv47noWWBvfwNbmolZU:b-g2lbq3) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L148)):*
> ```Python
> 
>         # Changes the ServerHardware power state
>         ov_client.server_hardware.update_power_state(
> ```

this should be in computer system. there should be a method to change this. All have shoud have to do here is instantiate a CS object and call change_power_state(new_power_state) where "change_power_state()" is a sample method on the computer system that would call the map method and call ov to change.
The problem is CS class needs OV connection. Need to discuss this.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 152 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4AM1xErFjqAFNftpL:-Kv4AM1yTcWiR_85l1z7:b-zcafpv) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L152)):*
> ```Python
> 
>         return Response(
>             response='{"ResetType": "%s"}' % reset_type,
> ```

Is this the correct response for DMTF when resetting a system?

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 164 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv48_DAQtz18xPYdIhi:-Kv48_DBZSudNdlsTqLh:b-wqlc4o) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L164)):*
> ```Python
>         # In case of error log exception and abort
>         logging.error('Mapping error: {}'.format(e))
>         abort(status.HTTP_400_BAD_REQUEST)
> ```

I think it should call 501 - not implemented

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 172 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv494cA14wZiIQVqabv:-Kv494cBYad5UTCyt3aR:bba13sb) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L172)):*
> ```Python
> 
> 
> @computer_system.errorhandler(status.HTTP_400_BAD_REQUEST)
> ```

if change above change here

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 175 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49RIIfKU6aTcb-6-A:-Kv49RIIfKU6aTcb-6-B:bdlxlx4) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L175)):*
> ```Python
> def bad_request(error):
>     """Creates a Bad Request Error response"""
>     logging.error(vars(error))
> ```

see below. remove it.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 185 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49H9s-vD22hUKALVZ:-Kv49H9tVltYQVgD4hCz:bdvcw8f) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L185)):*
> ```Python
> def not_found(error):
>     """Creates a Not Found Error response"""
>     logging.error(vars(error))
> ```

this should not be here. sorry my fault, was testing and didn't remove. Please remove all ocurences

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 42 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4Cxfr_PSm7XygKKYf:-Kv4Cxfr_PSm7XygKKYg:b-xbc53k) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L42)):*
> ```Python
>             - oneview unexpected exception
>             - know computer system
>     """
> ```

add the tests you are adding here

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 198 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49fuHAfHKW32EgMzB:-Kv49fuHAfHKW32EgMzC:brj8akg) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L198)):*
> ```Python
> 
>     @mock.patch.object(util, 'get_oneview_client')
>     def test_change_power_state(self, mock_get_ov_client):
> ```

add a comment telling us what this do

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 228 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49yOvqBee-lcExm-0:-Kv49yOvqBee-lcExm-1:b-tutm6r) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L228)):*
> ```Python
>             # Tests response
>             self.assertEqual(status.HTTP_200_OK, response.status_code)
>             self.assertEqual("application/json", response.mimetype)
> ```

you should test the retuning json to see if it worked IMHO

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 231 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4AVaJF-8mRQyCRzX4:-Kv4AVaK3By0gwje6Hfz:b1pilim) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L231)):*
> ```Python
> 
>     @mock.patch.object(util, 'get_oneview_client')
>     def test_change_power_state_invalid_value(self, mock_get_ov_client):
> ```

again, tell us what it should do, there's now way to check if it does simply based on method name. 

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 251 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4BAusU7bsTFhtYiDw:-Kv4BAusU7bsTFhtYiDx:b-28rbg4) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L251)):*
> ```Python
>         response = self.app.post("/redfish/v1/Systems/30303437-3034-4D32-3230"
>                                  "-313133364752/Actions/ComputerSystem.Reset",
>                                  data=dict(ResetType="PushPowerButon"))
> ```

Are you a testing an error with typo? (one t less?) Please be clear: ResetType="INVALID_TYPE" now I know what your doing. Was about the ask why should it fail when I noticed the typo. :(

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 260 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4CF7Em8iu96qyf0kR:-Kv4CF7Em8iu96qyf0kS:b-v9ftax) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L260)):*
> ```Python
>     def test_change_power_state_unexpected_error(self, mock_get_ov_client):
>         client = mock_get_ov_client()
>         client.server_hardware.get.side_effect = Exception()
> ```

this was already done in other tests. you should test your function, which is the one that changes de power. That should fail not the class building.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88)*
<!-- Sent from Reviewable.io -->

ffdarkpenguin commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88)*
<!-- Sent from Reviewable.io -->

llvieira commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 35 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv46Jok5qnmDzok2aKF:-Kv7dYJmu6_Vxw0tPjKw:b-bl9t1h) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L35)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

I don't agre with this. We removed url_prefix from app registering process because it there were conflict. If someone else implement /Systems/Whatever in another place will fail because you took the hole /Systems to you.
</blockquote></details>

Okay, removed.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 107 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv477SfYNIqZ0phW7tH:-Kv7dbbn56R3eZGNm6Mg:b-896fix) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L107)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

fix comment above and put the hole URI here. 
</blockquote></details>

Done.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 142 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv47ccPnmfQk_ECqSbi:-Kv7ddRhPtem13aHjV9e:bt2vr1b) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L142)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

its an object no a class
</blockquote></details>

Fixed.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 164 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv48_DAQtz18xPYdIhi:-Kv7eYw5wE6ofxqlfrfY:b-spg04y) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L164)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

I think it should call 501 - not implemented
</blockquote></details>

400 - The request could not be processed because it contains missing or invalid information (such as validation error on an input field, a missing required value, and so on). 
501 - The server does not (currently) support the functionality required to fulfill the request.

I think that if the user pass an invalid power state and it raise an error, this error it's not because the server does not support the functionality (501), but because there is invalid information (400).

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 172 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv494cA14wZiIQVqabv:-Kv7fPQyRu3ph0bsWjvA:b-43w9w9) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L172)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

if change above change here
</blockquote></details>

See what I said above...

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 175 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49RIIfKU6aTcb-6-A:-Kv7fVPV5aXr2d4ksIMP:bt2vr1b) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L175)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

see below. remove it.
</blockquote></details>

Fixed.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 185 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49H9s-vD22hUKALVZ:-Kv7fXWghKNZorc234Y0:bt2vr1b) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L185)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

this should not be here. sorry my fault, was testing and didn't remove. Please remove all ocurences
</blockquote></details>

Fixed.

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 42 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4Cxfr_PSm7XygKKYf:-Kv7gDK2LSjV1kfpfFJv:b-il4g1) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L42)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

add the tests you are adding here
</blockquote></details>

okay, done.

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 198 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49fuHAfHKW32EgMzB:-Kv7gdscmUtOhIIecca3:b-896fja) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L198)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

add a comment telling us what this do
</blockquote></details>

Done!

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 228 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49yOvqBee-lcExm-0:-Kv7hvWDpXdPBX5QTjW9:br2f382) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L228)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

you should test the retuning json to see if it worked IMHO
</blockquote></details>

Okay, done!

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 231 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4AVaJF-8mRQyCRzX4:-Kv7hxarSH-ypeldh8jy:b-896fja) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L231)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

again, tell us what it should do, there's now way to check if it does simply based on method name. 
</blockquote></details>

Done!

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 251 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4BAusU7bsTFhtYiDw:-Kv7hr6ZaWppH2k7KTUm:bt2vr1b) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L251)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

Are you a testing an error with typo? (one t less?) Please be clear: ResetType="INVALID_TYPE" now I know what your doing. Was about the ask why should it fail when I noticed the typo. :(
</blockquote></details>

Fixed.

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 260 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4CF7Em8iu96qyf0kR:-Kv7iwNGhJ0Hwc746Tql:b9dlee3) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L260)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

this was already done in other tests. you should test your function, which is the one that changes de power. That should fail not the class building.
</blockquote></details>

I do not understood. I do not have to try Unexpected Error in my function? Its two different functions and in each one can happen this type of error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88)*
<!-- Sent from Reviewable.io -->

llvieira commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 148 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv47noWWBvfwNbmolZT:-Kv7m-I-5G_lXsJbs29z:bn78ash) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L148)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

this should be in computer system. there should be a method to change this. All have shoud have to do here is instantiate a CS object and call change_power_state(new_power_state) where "change_power_state()" is a sample method on the computer system that would call the map method and call ov to change.
The problem is CS class needs OV connection. Need to discuss this.
</blockquote></details>

I guess that ComputerSystem class to know OneView connection it's not a good solution. Maybe an unnecessary dependence between this things... Really it's necessary to discuss this.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 152 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4AM1xErFjqAFNftpL:-Kv7lHHL18H4CxeTtJ_Z:bu2gh24) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L152)):*
<details><summary><i>Previously, ffdarkpenguin wrote…</i></summary><blockquote>

Is this the correct response for DMTF when resetting a system?
</blockquote></details>

http://redfish.dmtf.org/schemas/DSP0266_1.2.0.html#create-post-
If you see on that link, they show only the request body for the POST action, I don't find out the response body yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88)*
<!-- Sent from Reviewable.io -->

ffdarkpenguin commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 148 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv47noWWBvfwNbmolZT:-Kv8W07sUBo4XjqdmBBJ:b-w92kwz) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L148)):*
<details><summary><i>Previously, llvieira (Lucas Lima Vieira) wrote…</i></summary><blockquote>

I guess that ComputerSystem class to know OneView connection it's not a good solution. Maybe an unnecessary dependence between this things... Really it's necessary to discuss this.
</blockquote></details>

Lets wait for team opinion on this

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 152 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4AM1xErFjqAFNftpL:-Kv8hL9VEQMolbm-JKZV:b-pwmj0t) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L152)):*
<details><summary><i>Previously, llvieira (Lucas Lima Vieira) wrote…</i></summary><blockquote>

http://redfish.dmtf.org/schemas/DSP0266_1.2.0.html#create-post-
If you see on that link, they show only the request body for the POST action, I don't find out the response body yet.
</blockquote></details>

not found either, PO should look into that.

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 164 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv48_DAQtz18xPYdIhi:-Kv8hUYA4xz8HeTX99F4:b-87zsoy) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L164)):*
<details><summary><i>Previously, llvieira (Lucas Lima Vieira) wrote…</i></summary><blockquote>

400 - The request could not be processed because it contains missing or invalid information (such as validation error on an input field, a missing required value, and so on). 
501 - The server does not (currently) support the functionality required to fulfill the request.

I think that if the user pass an invalid power state and it raise an error, this error it's not because the server does not support the functionality (501), but because there is invalid information (400).
</blockquote></details>

Is this the case where user use something like reset_type="LUCAS" or the case where the user sent a valid redfish value but it not supported by oneview? If the first the error is 400 if the second the error is 501. The only way for you to know which is the case you should add a new error check: 
if reset_value is not a valid redfish value: error 400
if mapping failed: error 500

---

*[oneview_redfish_toolkit/blueprints/computer_system.py, line 172 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv494cA14wZiIQVqabv:-Kv8iGqSjcOey5uapCG_:b-6blvei) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/blueprints/computer_system.py#L172)):*
<details><summary><i>Previously, llvieira (Lucas Lima Vieira) wrote…</i></summary><blockquote>

See what I said above...
</blockquote></details>

see the reply above

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 198 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv49fuHAfHKW32EgMzB:-Kv8iZaMYD_0K7SBTzco:b-ivl046) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L198)):*
<details><summary><i>Previously, llvieira (Lucas Lima Vieira) wrote…</i></summary><blockquote>

Done!
</blockquote></details>

Please improve. You are not testing with a single value, you are testing with a range (all possible??) values. Tell about it here

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 260 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv4CF7Em8iu96qyf0kR:-Kv8j2LJIubSn_do5psM:blqzw25) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L260)):*
<details><summary><i>Previously, llvieira (Lucas Lima Vieira) wrote…</i></summary><blockquote>

I do not understood. I do not have to try Unexpected Error in my function? Its two different functions and in each one can happen this type of error.
</blockquote></details>

Tests for ov server_hardware error was already checked in the computer system class. no need for you to test it here. On the other hand you add a function to change the power state and NEVER make a test where that function gives you and error. That is what you should be testing.

---

*[oneview_redfish_toolkit/tests/blueprints/test_computer_system.py, line 281 at r1](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-Kv8jieSRcNzcnE0LV2B:-Kv8jieTuk2q4VQFLPQU:bwea218) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/caa88ed86d45e05a0103f2c4605442851d57e130/oneview_redfish_toolkit/tests/blueprints/test_computer_system.py#L281)):*
> ```Python
>         })
> 
>         client.server_hardware.get.side_effect = e
> ```

same as above here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88)*
<!-- Sent from Reviewable.io -->

danielreed commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[oneview_redfish_toolkit/api/computer_system.py, line 135 at r2](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88#-KvECC90F4jZnSWNzbC9:-KvECC90F4jZnSWNzbCA:b-g3qhss) ([raw file](https://github.com/hewlettpackard/oneview-redfish-toolkit/blob/3409db70a1d1c59451ff3cd53081d2dd3434876d/oneview_redfish_toolkit/api/computer_system.py#L135)):*
> ```Python
> 
>         reset_type_dict = dict()
> ```

Just a general comment on the mappings below, in OV when doing an 'On' operation you cannot also supply a powerControl value.  I believe powerControl value is only used for 'Off' operations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hewlettpackard/oneview-redfish-toolkit/88)*
<!-- Sent from Reviewable.io -->

ricardoas commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -109,3 +113,59 @@ def map_boot(self, boot_list):
+                reset_type is an unmapped value.
+        """
+
+        reset_type_dict = dict()
```
Can we build this dictionary statically and load it only once instead of each time this method is called?

llvieira replied:
Maybe a good solution would be I create a singleton class (ComputerSystemPower) to power.

llvieira replied:
Fixed.

ricardoas commented on oneview_redfish_toolkit/blueprints/computer_system.py:
```diff
@@ -103,6 +103,81 @@ def get_computer_system(uuid):
+            return abort(500)
+    """
+
+    reset_type = request.form["ResetType"]
```
Can this fail? What happens if someone sends it without the payload or even with a wrong name in the parameter?

llvieira replied:
Fixed.

ricardoas commented on oneview_redfish_toolkit/blueprints/computer_system.py:
```diff
@@ -103,6 +103,81 @@ def get_computer_system(uuid):
+        cs = ComputerSystem(sh, sht)
+
+        oneview_power_configuration = \
+            cs.get_oneview_power_configuration(reset_type)
```
What if someone calls it with a not allowed value?

llvieira replied:
This kind of thing is treated on get_oneview_power_configuration method.
```python
try:
       return reset_type_dict[reset_type]
except Exception:
       raise OneViewRedfishError(
                'There is no mapping for {} on the OneView'
                .format(reset_type))
```

ricardoas replied:
cool!

ricardoas commented on oneview_redfish_toolkit/tests/blueprints/test_computer_system.py:
```diff
@@ -193,3 +198,131 @@ def test_get_computer_system(self, mock_get_ov_client):
+        ov.server_hardware.get.return_value = sh_dict
+        ov.server_hardware_types.get.return_value = sht_dict
+
+        response = self.app.post("/redfish/v1/Systems/30303437-3034-4D32-3230"
```
I see you are testing the case I mentioned before... I'll take a look again!

llvieira replied:
Okay!

ricardoas commented on oneview_redfish_toolkit/tests/blueprints/test_computer_system.py:
```diff
@@ -193,3 +198,131 @@ def test_get_computer_system(self, mock_get_ov_client):
+        self.assertEqual("application/json", response.mimetype)
+
+    @mock.patch.object(util, 'get_oneview_client')
+    def test_change_power_state_unexpected_error(self, mock_get_ov_client):
```
How different are `test_change_power_state_unexpected_error` and `test_change_power_state_sh_exception`?

llvieira replied:
I think it's possible to occur a different Exception of HPOneViewException, so `test_change_power_state_unexpected_error` tests general Exception and `test_change_power_state_sh_exception` only the HPOneViewException.

ricardoas commented on oneview_redfish_toolkit/blueprints/computer_system.py:
```diff
@@ -103,6 +103,81 @@ def get_computer_system(uuid):
+            cs.get_oneview_power_configuration(reset_type)
+
+        # Changes the ServerHardware power state
+        ov_client.server_hardware.update_power_state(
```
Do we have a test for a failure in this? What happened to those cases of resetting a powered off node...

llvieira replied:
It's necessary to write a test for this case? If yes, I can do.

ffdarkpenguin commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -115,3 +143,39 @@ def map_boot(self, boot_list):
+
+        if reset_type == "PushPowerButton":
+            if self.server_hardware["powerState"] == "On":
+                POWER_STATE_MAP["PushPowerButton"]["powerState"] = "Off"
```
you should not change a CONSTANT.
Map it to a var and change the var, not the constant.

llvieira replied:
Fixed.

ffdarkpenguin commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -115,3 +143,39 @@ def map_boot(self, boot_list):
+                "errorCode": "NOT_IMPLEMENTED",
+                "message": "{} not mapped to OneView".format(reset_type)})
+
+        try:
```
you should do this at the beginning, if the map fails you don't have to do the rest

llvieira replied:
Fixed.

ffdarkpenguin replied:
Nice!

ffdarkpenguin commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -115,3 +143,39 @@ def map_boot(self, boot_list):
+            else:
+                POWER_STATE_MAP["PushPowerButton"]["powerState"] = "On"
+
+        if reset_type == "ForceOn" or reset_type == "Nmi":
```
Test abort conditions first, If you are going to abort why did you run the code above?!?

llvieira replied:
Fixed.

ffdarkpenguin commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -115,3 +143,39 @@ def map_boot(self, boot_list):
+            if self.server_hardware["powerState"] == "On":
+                POWER_STATE_MAP["PushPowerButton"]["powerState"] = "Off"
+            else:
+                POWER_STATE_MAP["PushPowerButton"]["powerState"] = "On"
```
again...

llvieira replied:
Fixed.

ricardoas commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -15,10 +15,36 @@
+        "powerControl": "MomentaryPress"
+    },
+    "GracefulRestart": {
+        "powerState": "On"
```
We really need powerControl = Reset here; oneview did not reset when I used this command... it continued with a powered on machine.

llvieira replied:
Fixed.

ricardoas commented on oneview_redfish_toolkit/api/computer_system.py:
```diff
@@ -15,10 +15,36 @@
+        "powerState": "On"
+    },
+    "ForceRestart": {
+        "powerState": "On"
```
The same as above but with powerControl = "ColdBoot"; we did discussed this in the channel... what happened?

llvieira replied:
Daniel said: "Just a general comment on the mappings below, in OV when doing an 'On' operation you cannot also supply a powerControl value. I believe powerControl value is only used for 'Off' operations.". So, I changed it. But I'll fixed it again. Thanks!

ricardoas replied:
I tried each combination in OneView and I saw that:
* when the SH is powered on, and you send `{ "powerState": "On"}` (redfish equivalent for `GracefulRestart`) it does not restart; but when you send `{ "powerState": "On", "powerControl": "Reset"}` it indeed restarted.
And the same happened with `{ "powerState": "On", "powerControl": "ColdBoot"}`.

Is there any other way of performing this reset operation, @danielreed?

danielreed replied:
My bad, sorry guys, Ricardo is right about those operations working if the power is already On.  It won't work if the power is Off.

ricardoas replied:
Yep! I tested this case too: with a powered off SH the reset operation fails as expected.

llvieira replied:
No problem.

ricardoas commented on oneview_redfish_toolkit/blueprints/computer_system.py:
```diff
@@ -105,6 +105,91 @@ def get_computer_system(uuid):
+
+    try:
+        try:
+            reset_type = request.form["ResetType"]
```
this line is supposing we are using a "form" type. According to the spec (http://redfish.dmtf.org/schemas/DSP0266_1.1.html#actions-property) is must be a application/json payload. I think it is available at `request.json["ResetType"]`
